resolwe/flow/managers/listener.py

```diff
@@ -0,0 +1,253 @@
+            obj['changeset'] = changeset
+            self.handle_update(obj)
+
+            try:
+                # Cleanup after processor
+                data_purge(data_ids=[obj['id']], delete=True, verbosity=self._verbosity)
+            except:  # pylint: disable=bare-except
+                logger.error(__("Purge error:\n\n{}", traceback.format_exc()))
+            # the executor will delete the key itself;
+            # since it was waiting for a reply for the finish command, deleting it here
+            # would lead to a race condition, and it would be bad to have too many timeouts here,
+            # waiting for the reply queue to drain
+
+    def handle_spawn(self, obj):
+        """Spawn new processes, as requested by the executor.
+
+        Command object format:
+        {
+            'action': 'spawn',
+            'id': [id of the Data object this command changes],
+            'cookie': [string value to be used as a reply identifier to this command],
+            'spawn_processors': [list of spawn dictionaries],
+            'exported_files_mapper': [dict with file mappings]
+        }
+        """
+        data_id = obj['id']
+        str_id = str(data_id)  # json only has string keys, so hydrate_spawned_files will need this stringified
+        exported_files_mapper = obj['exported_files_mapper']
+        with transaction.atomic():
+            parent_data = Data.objects.get(pk=data_id)
+
+            # Spawn processors
+            for d in obj['spawn_processors']:
+                d['contributor'] = parent_data.contributor
+                d['process'] = Process.objects.filter(slug=d['process']).latest()
+
+                for field_schema, fields in iterate_fields(d.get('input', {}), d['process'].input_schema):
+                    type_ = field_schema['type']
+                    name = field_schema['name']
+                    value = fields[name]
+
+                    if type_ == 'basic:file:':
+                        fields[name] = self.hydrate_spawned_files(exported_files_mapper, value, str_id)
+                    elif type_ == 'list:basic:file:':
+                        fields[name] = [self.hydrate_spawned_files(exported_files_mapper, fn, str_id) for fn in value]
+
+                with transaction.atomic():
+                    d = Data.objects.create(**d)
+                    d.parents.add(parent_data)
+
+                    # Copy permissions.
+                    copy_permissions(parent_data, d)
+
+                    # Copy collections.
+                    for collection in parent_data.collection_set.all():
+                        collection.data.add(d)
+
+        self._send_reply(obj, {'result': 'OK'})
+
+    def run(self):
+        """Start the main listening loop. Doesn't return until terminate() is called."""
+        # psycopg connection objects are thread-safe, but django makes them all thread-local,
+        # so we overwrite the new local connection with the one from the main thread,
```

Comment:
Hm, how is this synchronized? Because without proper synchronization, the other thread can put the database into an arbitrary state at an arbitrary point?

Reply:
Not applicable anymore.